vistautils/scripts/split_key_value_store.py

```diff
@@ -1,47 +1,121 @@
+*explicit_split* is useful for tasks which have standard train/test splits.
 """
 from contextlib import ExitStack
 
-from vistautils.io_utils import CharSink
-from vistautils.key_value import KeyValueSink, char_key_value_linear_source_from_params
+from immutablecollections import immutableset
+
+from vistautils.io_utils import CharSink, file_lines_to_set
+from vistautils.key_value import (
+    KeyValueSink,
+    KeyValueSource,
+    byte_key_value_linear_source_from_params,
+)
+from vistautils.misc_utils import str_list_limited
 from vistautils.parameters import Parameters
 from vistautils.parameters_only_entrypoint import parameters_only_entry_point
 
+_NUM_SLICES_PARAM = "num_slices"
+_EXPLICIT_SPLIT_PARAM = "explicit_split"
+
 
 def main(params: Parameters):
+    params.assert_exactly_one_present([_NUM_SLICES_PARAM, _EXPLICIT_SPLIT_PARAM])
+
+    with byte_key_value_linear_source_from_params(params) as input_source:
+        if _NUM_SLICES_PARAM in params:
+            _split_into_even_slices(input_source, params)
+        elif _EXPLICIT_SPLIT_PARAM in params:
+            _explicit_split(input_source, params)
+        else:
+            raise RuntimeError("No known split parameter specified.")
+
+
+def _split_into_even_slices(input_source: KeyValueSource[str, bytes], params: Parameters):
     output_directory = params.creatable_directory("output_dir")
     slices = params.positive_integer("num_slices")
-
     slice_paths = [output_directory / "{!s}.zip".format(i) for i in range(slices)]
     CharSink.to_file(output_directory / "_slices.txt").write(
         "\n".join(str(x) for x in slice_paths)
     )
-    output_sinks = [
-        KeyValueSink.zip_character_sink(slice_path) for slice_path in slice_paths
-    ]
-
+    output_sinks = [KeyValueSink.zip_bytes_sink(slice_path) for slice_path in slice_paths]
     # this is the magic incantation for handling variable-length lists of context managers
     with ExitStack() as exit_stack:
         for output_sink in output_sinks:
             exit_stack.enter_context(output_sink)
-        with char_key_value_linear_source_from_params(params) as input_source:
-            for (i, v) in enumerate(input_source.items()):
-                output_sinks[i % slices].put(v[0], v[1])
+        for (i, v) in enumerate(input_source.items()):
+            output_sinks[i % slices].put(v[0], v[1])
+
+
+def _explicit_split(source: KeyValueSource[str, bytes], params: Parameters):
+    explicit_split_namespace = params.namespace(_EXPLICIT_SPLIT_PARAM)
+
+    # We track these so we can ensure the split is a complete partition of the input,
+    # if the user so desires.
+    keys_copied = []
+
+    for split_namespace in explicit_split_namespace.sub_namespaces():
+        keys_for_split = file_lines_to_set(split_namespace.existing_file("keys_file"))
+        with KeyValueSink.zip_bytes_sink(
+            explicit_split_namespace.creatable_file("output_file")
```

Comment:
Shouldn't this also be
```suggestion
            split_namespace.creatable_file("output_file")
```